mail.google.com: Some gmail emails from financial institutions have weird box characters in them

vitorroriz opened the pull request:
#### af5873591243c10b2a035fa1e7991c8b98fe02bd
<pre>
mail.google.com: Some gmail emails from financial institutions have weird box characters in them
<a href="https://rdar.apple.com/137476880">rdar://137476880</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=285959">https://bugs.webkit.org/show_bug.cgi?id=285959</a>

Reviewed by Brent Fulgham.

ZWNJ should be rendered invisible

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/zwnj-renders-invisible-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/zwnj-renders-invisible-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/zwnj-renders-invisible.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fonts/ahem-visible-zwnj.otf: Added.

Canonical link: <a href="https://commits.webkit.org/289280@main">https://commits.webkit.org/289280@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b44867321dec7df531d9eeca0bc7dfd70d060d08

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86144 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5736 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40498 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91121 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37012 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88193 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5981 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13765 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66791 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24588 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89148 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78086 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47103 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32392 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36104 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33252 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92902 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13390 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9771 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75570 "Found 57 new test failures: compositing/blend-mode/non-separable-blend-modes.html compositing/layer-creation/will-change-layer-creation.html css3/filters/backdrop/backdrop-filter-does-not-size-properly-absolute.html editing/spelling/grammar-and-spelling-error-styling.html fast/css-custom-paint/out-of-memory-while-adding-worklet-module.html fast/dom/Window/open-window-min-size.html fast/inline/list-marker-inside-container-with-margin.html fast/inline/overflowing-content-with-hypens.html fast/multicol/columns-on-body.html fast/repaint/no-animation-outside-viewport-subframe.html ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13603 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73950 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74744 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18411 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18985 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17371 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13412 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18735 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13181 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16613 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14967 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -694,6 +694,10 @@ void ComplexTextController::adjustGlyphsAndAdvances()
+            } else if (character == zeroWidthNonJoiner) {
+                // zeroWidthNonJoiner is rendered as deletedGlyph for compatibility with other engines: https://bugs.webkit.org/show_bug.cgi?id=285959
+                advance.setWidth(0);
+                glyph = deletedGlyph;
```
Can we write a test case for this?

vitorroriz replied:
Yes. I intend to write a test for it, but I need to create a special Ahem font for it where that character is visible.

fantasai replied:
Should this catch only ZWNJ, or should this check be for all default-ignorable characters?

brentfulgham reviewed (approved):
Change looks good. I understand you can manually test this on a Gmail mail, but can you create a reduced test case we could use as a layout test so we don't break this in the future?

vitorroriz committed:
mail.google.com: Some gmail emails from financial institutions have weird box characters in them
rdar://137476880
https://bugs.webkit.org/show_bug.cgi?id=285959

Reviewed by Brent Fulgham.

ZWNJ should be rendered invisible

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/zwnj-renders-invisible-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/zwnj-renders-invisible-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/zwnj-renders-invisible.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fonts/ahem-visible-zwnj.otf: Added.

Canonical link: https://commits.webkit.org/289280@main
